notebook/exploration-reports/2022.03-ipld-url-scheme.md

```diff
@@ -0,0 +1,106 @@
+```
+
+## Objective
+
+`ipfs://` exists out in the wild to provide ways to link to files, and some web browsers integrate IPFS URLs to load and author content.
+This spec shows how a similar URL scheme can exist for IPLD data.
+Part of the goal of the spec is to make it easy to reuse more or less the same APIs between software that handles `ipld://` URLs natively, HTTP gateways like the [IPFS gateway](https://docs.ipfs.io/concepts/ipfs-gateway/) and CLIs.
+With that in mind, there should be a clear mapping between HTTP-like APIs via gateways and URL handlers, and CLI arguments.
+In order to make it easier to come to an initial spec, we will focus on a minimal set of features and make subsequent specs with support for features like ADLs, Schemas, and advanced selectors.
+
+## Initial Functionality
+
+The initial functionality focuses on being able to GET and PUT IPLD data to and from a URL.
+For now, this will be done by encoding individual blocks of IPLD data using codecs with a default to the [DAG-JSON](https://ipld.io/specs/codecs/dag-json/spec/) codec.
+Some initial support for pathing is also included for convenience.
+
+### GET `ipld://{CID}/{Path} Headers {Accept="application/vnd.ipld.dag-json"} => Data`
+
+The most basic mapping is to get some data from a CID.
+
+The `CID` parameter in the URL MUST be in the CIDv1 format or any other format that is not case sensitive.
+This is necessary since the URL scheme considers the `hostname` portion to be case-insensitive and many URL parsers (such as the one in Chrome) will automatically convert it to lowercase. 
+
+`Data` is the response body which contaisn the encoded IPLD data for the given CID+Path.
+Note that any `Link`s within the IPLD data will not be resolved, and you'll need to conver them to URLs and fetch additional data yourself, or add path segments to the URL to traverse into them.
+
+The `Accept` header may be optionally provided to specify the codec that should be used in the response.
+This follows the same conventions as in the [IPFS gateway](https://github.com/ipfs/go-ipfs/pull/8758).
+Other codecs that may be targetted are DAG-CBOR.
+TODO: `application/vnd.ipld.dag-json` or just `application/json`?
+
+The `Path` can be used to specify segments separated by a `/` to use with basic traversal,
+if it is supplied the IPLD Node at that location relative to the CID will be returned.
+If there path does not lead to a valid IPLD node, an error will be returned instead, e.g. `HTTP 404`.
+
+### PUT `ipld://{CID}/{path} Headers {Content-Type="application/vnd.ipld.dag-json"} Body={encoded IPLD data} => URL`
+
+The `{CID}` should be the "base" CID to extend data from.
+Sadly if we are to use URLs (TODO: why should we use URLs?), we must have some sort of CID in the hostname.
+TODO: Document some inline CIDs to use. IPFS has this empty directory CID `ipfs://bafyaabakaieac` which would be cool to have analogs for for empty `{}` object and `[]` array nodes.
```

Comment:
Well, we now have the concept of "unit" type in schema-land: https://github.com/ipld/ipld/blob/cfe4b4914b6c5a6e37c5cc9741528e0022337f98/specs/schemas/schema-schema.ipldsch#L761-L789

And we have identity CIDs, we could just pick one to standardise on as a base.

https://cid.ipfs.io/#baeaaaapw might be a good choice, `null` is a commonly understood unit value and might be a good "there's nothing here to start from".

Or are you wanting to convey codec information with the "base"? So, Content-Type tells the receiver what format the data is being provided in, but the base CID says what codec should be used for the encoded block? Do we want that kind of flexibility? And if so, what about base CIDs that have some data because they're CIDs that reference some existing block or have some data in them as identity CIDs?

Yet another possibility here is to just use identity CIDs to send the data and `Content-Type` to say what codec to encode into—what limitations do we run into with hostnames here?

Reply:
Also, maybe the semantics of a `PUT` could all fall under IPLD Patch since it already has all needed modification primitives (and more)? Otherwise this would be introducing a mutation operation that is limited to only adding or changing data from existing blocks, whereas Patch gives us full modification potential.

Reply:
> Or are you wanting to convey codec information with the "base"?

I think keeping the codec information in the `Content-Type` header would be best to keep things URL-y. `ipld://baeaaaapw/` seems like a great place to start.

> what about base CIDs that have some data because they're CIDs that reference some existing block or have some data in them as identity CIDs?

I think this would be useful. e.g. It'd be cool to have some base CIDs with examples like an empty array or an empty map. Is there an existing list I could reference somewhere? Should I just generate these myself with the ipld CLI?
'

> maybe the semantics of a PUT could all fall under IPLD Patch since it already has all needed modification primitives

I was actually thinking that the `PATCH` HTTP method could be used there instead. I think it'd be cool to add support for it as a first-class feature, but since Patch is still a bit of a WIP I didn't want to block this to wait on it.

One thing to consider here is that we just need to have _something_ in the hostname portion of the URL since in my experience most parsers (like browsers) will freak out if it's only _sometimes_ there. 😅

Alternately, in IPNS land we were looking at using `localhost` as a "no-op" name for when we want to do something with data that isn't already tied to a particular CID. @lidel might have more to say on that front.

Reply:
- I think my idea was that if we want to invent some name to indicate "empty" state, the `localhost`  is not a bad choice because it is something devs understand, resolves safely to loopback device if DNS is ever involved (either on purpose, or as a bug),  and also hints that data is added to local IPLD node, and not some magical cloud:
  - `PUT ipld://localhost`  → get CID

- In case a drive-by reader is confused by "JSON Patch", it is different  thing from "HTTP PATCH" method:
   - The "JSON Patch"   @rvagg mentioned is a convention ([RFC 6902](http://tools.ietf.org/html/rfc6902) from the IETF) for representing JSON diffs (http://jsonpatch.com) – see relevant example in https://github.com/ipfs/go-ipfs/issues/4801#issuecomment-1024303462  (and the discussion that follows).

Reply:
I'm into having `ipld://localhost/` be a special thing for when there's no base CID to build off of. @rvagg @warpfork Any strong feelings on that matter?

Regarding PATCH, I've added a link to the GH issue where the implementation is being tracked: https://github.com/ipld/go-ipld-prime/pull/350

Reply:
seems reasonable to me; it is slightly arbitrary (e.g. why not `::` to be uber-modern?) but I guess anything we choose is!